docs/RFCS/20180413_alter_primary_key.md

```diff
@@ -41,183 +29,273 @@ operational considerations as well.
-the old PK index stays around as a unique index. Expand this to cover what
-follows.
-
-### Special Handling of Primary Indexes
-
-In various places, the descriptor handling of the primary index is special-cased
-with respect to handling stored or extra columns. With the ability to swap the
-index which is "primary", such special-handling will need to be revised. Ideally
-it can be simply removed, e.g. all columns just appear in the extra or stored
-lists, rather than being implicitly inferred to be there if primary.  In many
-cases removing special handling may simplify some code that currently needs to
-expand these implicit cases.
-
-Beyond the mechanical metadata computation and bookkeeping, there are some cases
-where the primary index currently has special-case behaviors, chiefly support
-for column families, which will need to be expanded other indexes as well.
-
-### TODO: Examples (descriptors before and after)
+Now we are moving to a global deployment, so we would like to add a new
+`region` column which is part of the primary key. First we add the column,
+which must not be nullable:
+```sql
+ALTER TABLE users ADD COLUMN region STRING NOT NULL;
+```
 
-### Implementation Plan
+Next, we change the primary key.
+```sql
+ALTER TABLE users ALTER PRIMARY KEY (region, id);
+```
 
-Implementing the above likely begins by eliminating as much special-handling of
-the primary index as possible: adding family support to all indexes, adding
-explicit fields that track columns stored, implicitly indexed, etc and/or always
-populating them and using them for all indexes, primary or otherwise. Once the
-primary index is no longer given special treatment, the swap above, when its
-conditions are met, should be possible.
+The old primary index has been rewritten as a secondary index named
+`user_id_idx`. It can be dropped it if no longer needed.
+```sql
+DROP INDEX users@users_id_idx;
+```
 
-### Follow-up, Out-of-Scope Work
+## Rewriting indexes
+
+While simple at a high level, there are several key technical details involved
+in applying the steps above. Consider again the example of prepending a
+`region` column to the `users` table's primary key. The steps we want to
+perform behind the scenes are:
+1. Build a new primary index on `(region, id)`
+2. Build a new secondary index `users_name_idx_rewritten` that uses the new
+   primary index's columns for key uniqueness.
+3. Build a new secondary index `users_id_idx` which indexes the same columns as
+   the old primary key.
+4. Swap the old primary index with the new primary index and `users_name_idx`
+   with `users_name_idx_rewritten`
+5. Delete the old `users_name_idx` and the old primary key.
+
+There are several reasons we cannot accomplish these steps using existing SQL
+syntax.
+1. The new primary index needs to be *built and encoded as a primary index*. We
+   cannot build it as a secondary index and suddenly switch to using it as a
+   primary index.
```

Comment:
If the new PK contains the old PK (typical if we are prefixing it with a region), and we build the new PK as a secondary index against the old PK, is there a difference between primary and secondary index encoding? (assuming secondary encoding abides by the column families).

Reply:
In this case, the value encoding would be slightly different (the secondary encoding uses the BYTES value tag for its values, and the primary encoding would use the TUPLE value tag). Places where we do CPUT's would run into issues in that case.